File: Ch9/02.smsh/smsh1-rs/src/execute.rs
use std::process::{Command, exit};
use std::os::unix::process::CommandExt;

use nix::unistd::{fork, ForkResult};
use nix::sys::signal::{signal, SigHandler, SIGINT, SIGQUIT};
use nix::sys::wait::{wait, WaitStatus};


pub fn execute(mut cmd: Command) -> Option<WaitStatus> {
    match unsafe{fork()} {
        Ok(ForkResult::Child) => {
            // enable default signal handling for SIGINT and SIGQUIT
            unsafe{
                let _ = signal(SIGINT, SigHandler::SigDfl);
                let _ = signal(SIGQUIT, SigHandler::SigDfl);
            }
            eprintln!("{}", cmd.exec());
            exit(-1);
        },
        Ok(ForkResult::Parent { child: _ }) => {
            return wait().ok(); 
        },
        Err(msg) => {
            eprintln!("fork() error: {:?}", msg);
            exit(-1);
        },
    }
}

File: Ch9/02.smsh/smsh3-c/execute.c
/*
    execute.c
    code used by small shell to eecute commands
*/

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <sys/wait.h>
#include <assert.h>
#include <unistd.h>

/*
    purpose: run a program passing its arguments
    returns status retutned via wait, or -1 on error
    errors: -1 on fork() or wait() errors
*/
int execute(char *argv[]) {
    pid_t pid;
    int child_info = -1;


    if (NULL == argv[0]) {
        return 0;
    }

    if (-1 == (pid = fork())) {
        perror("fork()");
        return -1;
    } else if (0 == pid) {
        // enable default signal handling for SIGINT and SIGQUIT        
        signal(SIGINT, SIG_DFL);
        signal(SIGQUIT, SIG_DFL);

        // execute the command
        if (-1 == execvp(argv[0], argv)) {
            perror("execvp()");
            exit(-1);
        }
    } else {
        if (-1 == wait(&child_info)) {
            perror("wait()");
            return -1;
        }
    }

    return child_info;
}